File: rust/e6.rs

/*
Finds the difference between the sum of the squares of the first one hundred
natural numbers and the square of the sum.

NOTE: the problem ACTUALLY wants the difference between the square of the sum
and the sum of the squares.
*/

fn euler_6_easy_way() -> usize {
    let sum_of_squares: usize = (1..=100).map(|x: usize| x.pow(2)).sum();
    let square_of_sum = (1..=100).sum::<usize>().pow(2);
    square_of_sum - sum_of_squares
}

fn euler_6_better_way() -> usize {
    //uses S_n = n(a_1 + a_n)/2
    //to be honest, I forget how sum_of_squares works...
    //TODO: sum_of_squares needs to be calculated as a float then cast to int
    //..I think.  Probably a better way.
    let sum_of_squares = (1_000_000/3) + (10_000/2) + (100/6); //338350
    let square_of_sum = (100*101/2 as usize).pow(2);
    square_of_sum - sum_of_squares
}

fn main() {
    println!("easy way: {}", euler_6_easy_way());
    println!("better way: {}", euler_6_better_way());
}

File: python/e6.py

"""
Finds the difference between the sum of the squares of the first one hundred
natural numbers and the square of the sum.

NOTE: the problem ACTUALLY wants the difference between the square of the sum
and the sum of the squares.
"""

def Euler_6_easy_way():
    sum_of_squares = sum([(i**2) for i in range (1,101)])
    square_of_sum = (sum([i for i in range (1,101)]))**2
    return square_of_sum - sum_of_squares

def Euler_6_better_way():
    # uses S_n = n(a_1 + a_n)/2 for square of sums
    # to be honest... I can't quite remember how sum_of_squares works.  But it does.
    sum_of_squares = int((1e+6)/3 + (1e+4)/2 + 100/6)
    square_of_sum = int((100*101/2)**2)
    return square_of_sum - sum_of_squares

if __name__ == "__main__":
    print(f"easy way: {Euler_6_easy_way()}")
    print(f"better way: {Euler_6_better_way()}")
